shell/platform/android/io/flutter/embedding/engine/android/AndroidTouchProcessor.java

```diff
@@ -0,0 +1,186 @@
+  private static final int POINTER_CHANGE_HOVER = 3;
+  private static final int POINTER_CHANGE_DOWN = 4;
+  private static final int POINTER_CHANGE_MOVE = 5;
+  private static final int POINTER_CHANGE_UP = 6;
+
+  // Must match the PointerDeviceKind enum in pointer.dart.
+  // TODO(mattcarroll): convert these to an IntDef
+  private static final int POINTER_DEVICE_KIND_TOUCH = 0;
+  private static final int POINTER_DEVICE_KIND_MOUSE = 1;
+  private static final int POINTER_DEVICE_KIND_STYLUS = 2;
+  private static final int POINTER_DEVICE_KIND_INVERTED_STYLUS = 3;
+  private static final int POINTER_DEVICE_KIND_UNKNOWN = 4;
+
+  // Must match the unpacking code in hooks.dart.
+  private static final int POINTER_DATA_FIELD_COUNT = 19;
+  private static final int BYTE_PER_FIELD = 8;
+
+  @NonNull
+  private final FlutterRenderer renderer;
+
+  /**
+   * Constructs an {@code AndroidTouchProcessor} that will send touch event data
+   * to the Flutter execution context represented by the given {@link FlutterRenderer}.
+   */
+  // TODO(mattcarroll): consider moving packet behavior to a FlutterInteractionSurface instead of FlutterRenderer
+  public AndroidTouchProcessor(@NonNull FlutterRenderer renderer) {
+    this.renderer = renderer;
+  }
+
+  /**
+   * Sends the given {@link MotionEvent} data to Flutter in a format that
+   * Flutter understands.
+   */
+  public boolean onTouchEvent(MotionEvent event) {
+    int pointerCount = event.getPointerCount();
+
+    // Prepare a data packet of the appropriate size and order.
+    ByteBuffer packet = ByteBuffer.allocateDirect(
+        pointerCount * POINTER_DATA_FIELD_COUNT * BYTE_PER_FIELD
+    );
+    packet.order(ByteOrder.LITTLE_ENDIAN);
+
+    int maskedAction = event.getActionMasked();
+    // ACTION_UP, ACTION_POINTER_UP, ACTION_DOWN, and ACTION_POINTER_DOWN
+    // only apply to a single pointer, other events apply to all pointers.
+    if (maskedAction == MotionEvent.ACTION_UP || maskedAction == MotionEvent.ACTION_POINTER_UP
+        || maskedAction == MotionEvent.ACTION_DOWN || maskedAction == MotionEvent.ACTION_POINTER_DOWN) {
+      addPointerForIndex(event, event.getActionIndex(), packet);
+    } else {
+      // ACTION_MOVE may not actually mean all pointers have moved
+      // but it's the responsibility of a later part of the system to
+      // ignore 0-deltas if desired.
+      for (int p = 0; p < pointerCount; p++) {
+        addPointerForIndex(event, p, packet);
+      }
+    }
+
+    // Verify that the packet is the expected size.
+    assert packet.position() % (POINTER_DATA_FIELD_COUNT * BYTE_PER_FIELD) == 0;
+
+    // Send the packet to Flutter.
+    renderer.dispatchPointerDataPacket(packet, packet.position());
+
+    return true;
+  }
+
+  // TODO(mattcarroll): consider creating a PointerPacket class instead of using a procedure that mutates inputs.
+  private void addPointerForIndex(MotionEvent event, int pointerIndex, ByteBuffer packet) {
```

Comment:
If there's documentation for how we're organizing these packets, it should be linked here.

If there isn't documentation for how we're organizing these packets, you should write it while you still remember how this works.